File: Sources/ScrcapCore/KeymapEngine.swift
// KeymapEngine — portable core (P4). Action ↔ chord mapping, chord parsing,
// conflict detection. Keys are named abstractly ("2", "r", "f5"); translation
// to OS keycodes (Carbon on macOS) lives in the platform layer.

import Foundation

public struct ChordModifiers: OptionSet, Codable, Hashable, Sendable {
    public let rawValue: UInt8
    public init(rawValue: UInt8) { self.rawValue = rawValue }

    public static let command = ChordModifiers(rawValue: 1 << 0)
    public static let option  = ChordModifiers(rawValue: 1 << 1)
    public static let shift   = ChordModifiers(rawValue: 1 << 2)
    public static let control = ChordModifiers(rawValue: 1 << 3)
}

public struct KeyChord: Codable, Hashable, Sendable {
    /// Normalized lowercase key name: "a"…"z", "0"…"9", "f1"…"f12",
    /// "space", "esc", punctuation characters.
    public var key: String
    public var modifiers: ChordModifiers

    public init(key: String, modifiers: ChordModifiers) {
        self.key = key.lowercased()
        self.modifiers = modifiers
    }

    /// Parses "opt+shift+2", "cmd+shift+z", "ctrl+alt+space". Returns nil on
    /// empty key or unknown modifier token.
    public init?(string: String) {
        var mods: ChordModifiers = []
        var key: String?
        for raw in string.split(separator: "+") {
            let token = raw.trimmingCharacters(in: .whitespaces).lowercased()
            switch token {
            case "cmd", "command", "⌘": mods.insert(.command)
            case "opt", "option", "alt", "⌥": mods.insert(.option)
            case "shift", "⇧": mods.insert(.shift)
            case "ctrl", "control", "⌃": mods.insert(.control)
            case "": continue
            default:
                if key != nil { return nil } // two non-modifier tokens
                key = token
            }
        }
        guard let k = key, !k.isEmpty else { return nil }
        self.init(key: k, modifiers: mods)
    }

    /// Canonical storage form, e.g. "opt+shift+2".
    public var stringValue: String {
        var parts: [String] = []
        if modifiers.contains(.control) { parts.append("ctrl") }
        if modifiers.contains(.option) { parts.append("opt") }
        if modifiers.contains(.shift) { parts.append("shift") }
        if modifiers.contains(.command) { parts.append("cmd") }
        parts.append(key)
        return parts.joined(separator: "+")
    }

    /// Display form with mac symbols, e.g. "⌥⇧2".
    public var displayValue: String {
        var s = ""
        if modifiers.contains(.control) { s += "⌃" }
        if modifiers.contains(.option) { s += "⌥" }
        if modifiers.contains(.shift) { s += "⇧" }
        if modifiers.contains(.command) { s += "⌘" }
        return s + key.uppercased()
    }
}

/// Globally-hotkeyed actions. Editor-local keys (A/R/C, 1–7, Esc…) are
/// handled by the editor against `Settings.editorKeys`.
public enum AppAction: String, Codable, Sendable {
    case captureRegion
    case captureWindow
    case captureFullscreen
    case captureScrolling
    case repeatLast

    public static let shortcutOrder: [AppAction] = [
        .captureRegion, .captureWindow, .captureFullscreen, .captureScrolling, .repeatLast,
    ]

    public var title: String {
        switch self {
        case .captureFullscreen: return "Capture Fullscreen"
        case .captureRegion: return "Capture Region"
        case .captureWindow: return "Capture Window"
        case .captureScrolling: return "Scrolling Capture"
        case .repeatLast: return "Repeat Last Capture"
        }
    }
}

public struct Keymap: Codable, Equatable, Sendable {
    public private(set) var bindings: [AppAction: KeyChord]

    /// Region is the daily-driver capture, so it gets the primary slot (⌥⇧1).
    public static let defaults = Keymap(bindings: [
        .captureRegion: KeyChord(key: "1", modifiers: [.option, .shift]),
        .captureWindow: KeyChord(key: "2", modifiers: [.option, .shift]),
        .captureFullscreen: KeyChord(key: "3", modifiers: [.option, .shift]),
        .captureScrolling: KeyChord(key: "4", modifiers: [.option, .shift]),
        .repeatLast: KeyChord(key: "r", modifiers: [.option, .shift]),
    ])

    public init(bindings: [AppAction: KeyChord]) {
        self.bindings = bindings
    }

    public func chord(for action: AppAction) -> KeyChord? { bindings[action] }

    /// The action already bound to `chord`, if any (excluding `excluding`).
    public func conflict(for chord: KeyChord, excluding: AppAction? = nil) -> AppAction? {
        bindings.first { $0.value == chord && $0.key != excluding }?.key
    }

    /// Rebinds `action`. Returns the action the chord was stolen from, if the
    /// chord was already in use (the caller decides how to surface that).
    @discardableResult
    public mutating func set(_ chord: KeyChord, for action: AppAction) -> AppAction? {
        let conflicted = conflict(for: chord, excluding: action)
        if let conflicted { bindings.removeValue(forKey: conflicted) }
        bindings[action] = chord
        return conflicted
    }

    /// Chords claimed by macOS itself that we warn about in the recorder UI.
    public static let systemReserved: Set<KeyChord> = [
        KeyChord(key: "3", modifiers: [.command, .shift]),
        KeyChord(key: "4", modifiers: [.command, .shift]),
        KeyChord(key: "5", modifiers: [.command, .shift]),
        KeyChord(key: "space", modifiers: [.command]),
        KeyChord(key: "q", modifiers: [.command]),
        KeyChord(key: "tab", modifiers: [.command]),
    ]

    public static func isSystemReserved(_ chord: KeyChord) -> Bool {
        systemReserved.contains(chord)
    }
}

File: scrcap-windows/src/Scrcap.Core/KeymapEngine.cs
namespace Scrcap.Core;

[Flags]
public enum ChordModifiers : byte
{
    None = 0,
    Command = 1 << 0,
    Option = 1 << 1,
    Shift = 1 << 2,
    Control = 1 << 3,
}

public readonly record struct KeyChord(string Key, ChordModifiers Modifiers)
{
    public string Key { get; } = Key.ToLowerInvariant();

    public static bool TryParse(string value, out KeyChord chord)
    {
        chord = default;
        ChordModifiers modifiers = ChordModifiers.None;
        string? key = null;

        foreach (var raw in value.Split('+', StringSplitOptions.TrimEntries))
        {
            var token = raw.ToLowerInvariant();
            switch (token)
            {
                case "cmd":
                case "command":
                case "win":
                case "windows":
                case "meta":
                case "⌘":
                    modifiers |= ChordModifiers.Command;
                    break;
                case "opt":
                case "option":
                case "alt":
                case "⌥":
                    modifiers |= ChordModifiers.Option;
                    break;
                case "shift":
                case "⇧":
                    modifiers |= ChordModifiers.Shift;
                    break;
                case "ctrl":
                case "control":
                case "⌃":
                    modifiers |= ChordModifiers.Control;
                    break;
                case "":
                    break;
                default:
                    if (key is not null)
                    {
                        return false;
                    }

                    key = token;
                    break;
            }
        }

        if (string.IsNullOrWhiteSpace(key))
        {
            return false;
        }

        chord = new KeyChord(key, modifiers);
        return true;
    }

    public string StringValue
    {
        get
        {
            var parts = new List<string>();
            if (Modifiers.HasFlag(ChordModifiers.Control))
            {
                parts.Add("ctrl");
            }

            if (Modifiers.HasFlag(ChordModifiers.Option))
            {
                parts.Add("opt");
            }

            if (Modifiers.HasFlag(ChordModifiers.Shift))
            {
                parts.Add("shift");
            }

            if (Modifiers.HasFlag(ChordModifiers.Command))
            {
                parts.Add("cmd");
            }

            parts.Add(Key);
            return string.Join('+', parts);
        }
    }

    public string WindowsDisplayValue
    {
        get
        {
            var parts = new List<string>();
            if (Modifiers.HasFlag(ChordModifiers.Control))
            {
                parts.Add("Ctrl");
            }

            if (Modifiers.HasFlag(ChordModifiers.Option))
            {
                parts.Add("Alt");
            }

            if (Modifiers.HasFlag(ChordModifiers.Shift))
            {
                parts.Add("Shift");
            }

            if (Modifiers.HasFlag(ChordModifiers.Command))
            {
                parts.Add("Win");
            }

            parts.Add(Key.ToUpperInvariant());
            return string.Join('+', parts);
        }
    }
}

public enum AppAction
{
    CaptureRegion,
    CaptureWindow,
    CaptureFullscreen,
    CaptureScrolling,
    CaptureDelayed,
    RepeatLast,
}

public static class AppActionExtensions
{
    public static IReadOnlyList<AppAction> ShortcutOrder { get; } =
    [
        AppAction.CaptureRegion,
        AppAction.CaptureWindow,
        AppAction.CaptureFullscreen,
        AppAction.CaptureScrolling,
        AppAction.CaptureDelayed,
        AppAction.RepeatLast,
    ];

    public static string StorageKey(this AppAction action) =>
        action switch
        {
            AppAction.CaptureRegion => "captureRegion",
            AppAction.CaptureWindow => "captureWindow",
            AppAction.CaptureFullscreen => "captureFullscreen",
            AppAction.CaptureScrolling => "captureScrolling",
            AppAction.CaptureDelayed => "captureDelayed",
            AppAction.RepeatLast => "repeatLast",
            _ => throw new ArgumentOutOfRangeException(nameof(action), action, null),
        };

    public static bool TryParseStorageKey(string value, out AppAction action)
    {
        action = value switch
        {
            "captureRegion" => AppAction.CaptureRegion,
            "captureWindow" => AppAction.CaptureWindow,
            "captureFullscreen" => AppAction.CaptureFullscreen,
            "captureScrolling" => AppAction.CaptureScrolling,
            "captureDelayed" => AppAction.CaptureDelayed,
            "repeatLast" => AppAction.RepeatLast,
            _ => default,
        };

        return value is "captureRegion" or "captureWindow" or "captureFullscreen" or "captureScrolling" or "captureDelayed" or "repeatLast";
    }

    public static string Title(this AppAction action) =>
        action switch
        {
            AppAction.CaptureFullscreen => "Capture Fullscreen",
            AppAction.CaptureRegion => "Capture Region",
            AppAction.CaptureWindow => "Capture Window",
            AppAction.CaptureScrolling => "Scrolling Capture",
            AppAction.CaptureDelayed => "Delayed Region Capture",
            AppAction.RepeatLast => "Repeat Last Capture",
            _ => throw new ArgumentOutOfRangeException(nameof(action), action, null),
        };
}

public sealed class Keymap
{
    public Keymap(IDictionary<AppAction, KeyChord> bindings)
    {
        Bindings = new Dictionary<AppAction, KeyChord>(bindings);
    }

    public static Keymap Defaults { get; } = new(new Dictionary<AppAction, KeyChord>
    {
        [AppAction.CaptureRegion] = new("1", ChordModifiers.Option | ChordModifiers.Shift),
        [AppAction.CaptureWindow] = new("2", ChordModifiers.Option | ChordModifiers.Shift),
        [AppAction.CaptureFullscreen] = new("3", ChordModifiers.Option | ChordModifiers.Shift),
        [AppAction.CaptureScrolling] = new("4", ChordModifiers.Option | ChordModifiers.Shift),
        [AppAction.CaptureDelayed] = new("5", ChordModifiers.Option | ChordModifiers.Shift),
        [AppAction.RepeatLast] = new("r", ChordModifiers.Option | ChordModifiers.Shift),
    });

    public Dictionary<AppAction, KeyChord> Bindings { get; }

    public KeyChord? ChordFor(AppAction action) =>
        Bindings.TryGetValue(action, out var chord) ? chord : null;

    public AppAction? ConflictFor(KeyChord chord, AppAction? excluding = null)
    {
        foreach (var (action, existing) in Bindings)
        {
            if (existing == chord && action != excluding)
            {
                return action;
            }
        }

        return null;
    }

    public AppAction? Set(KeyChord chord, AppAction action)
    {
        var conflicted = ConflictFor(chord, action);
        if (conflicted is { } victim)
        {
            Bindings.Remove(victim);
        }

        Bindings[action] = chord;
        return conflicted;
    }

    public static IReadOnlySet<KeyChord> SystemReserved { get; } = new HashSet<KeyChord>
    {
        new("l", ChordModifiers.Command),
        new("d", ChordModifiers.Command),
        new("tab", ChordModifiers.Command),
        new("f4", ChordModifiers.Option),
        new("delete", ChordModifiers.Control | ChordModifiers.Option),
    };

    public static bool IsSystemReserved(KeyChord chord) => SystemReserved.Contains(chord);

    public static bool IsUsableGlobalShortcut(KeyChord chord) =>
        chord.Modifiers != ChordModifiers.None && !IsSystemReserved(chord);
}
